src/k2/dto/K23SIInspect.h

```diff
@@ -0,0 +1,97 @@
+
+    The above copyright notice and this permission notice shall be included in all copies
+    or
+    substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED "AS IS",
+    WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+    DAMAGES OR OTHER
+    LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+    SOFTWARE.
+*/
+
+#pragma once
+
+#include <k2/common/Common.h>
+#include <k2/module/k23si/TxnManager.h>
+#include <k2/transport/PayloadSerialization.h>
+#include <k2/transport/Status.h>
+
+namespace k2 {
+namespace dto {
+
+// For test and debug, requests all versions including WIs for a particular key
+// without affecting transaction state
+struct K23SIInspectRecordsRequest {
+    // These fields make the request compatible with the PartitionRequest wrapper
+    Partition::PVID pvid; // the partition version ID. Should be coming from an up-to-date partition map
+    String collectionName;
+    Key key; // the key to gather all records for
+    K2_PAYLOAD_FIELDS(pvid, collectionName, key);
+};
+
+struct K23SIInspectRecordsResponse {
+    std::vector<DataRecord> records;
+    K2_PAYLOAD_FIELDS(records);
+};
+
+// For test and debug, requests the TRH of a transaction without affecting transaction state
+struct K23SIInspectTxnRequest {
+    // These fields make the request compatible with the PartitionRequest wrapper
+    Partition::PVID pvid; // the partition version ID. Should be coming from an up-to-date partition map
+    String collectionName;
+    Key key; // the key of the THR to request
+    K23SI_MTR mtr;
+    K2_PAYLOAD_FIELDS(pvid, collectionName, key, mtr);
+};
+
+// Contains the TRH data (struct TxnRecord) without the internal
+// management members such as intrusive list hooks
+struct K23SIInspectTxnResponse {
+    TxnId txnId;
+
+    // the keys to which this transaction wrote. These are delivered as part of the End request and we have to ensure
+    // that the corresponding write intents are converted appropriately
+    std::vector<dto::Key> writeKeys;
+
+    // Expiry time point for retention window - these are driven off each TSO clock update
+    dto::Timestamp rwExpiry;
+
+    bool syncFinalize = false;
+
+    TxnRecord::State state;
+
+    K2_PAYLOAD_FIELDS(txnId, writeKeys, rwExpiry, state);
+
+    friend std::ostream& operator<<(std::ostream& os, const K23SIInspectTxnResponse& rec) {
+        os << "{txnId=" << rec.txnId << ", writeKeys=[";
+        os << rec.writeKeys.size();
+        os << "], rwExpiry=" << rec.rwExpiry << ", syncfin=" << rec.syncFinalize << "}";
+        return os;
+    }
+};
+
+// For test and debug, requests all WIs on a node for any key
```

Comment:
is this different that inspectRecords?

Reply:
The comment should say "all keys". The idea is to use it for the orphaned WI scan.

Reply:
so i guess we should have InspectKey (all versions for 1 key) and InspectAllKeys(all versions for all keys) ?

Reply:
I added InspectAllKeys, but it only returns the keys themselves. I figured that returning all the versions too would be too big for one request. Sound good?

Reply:
makes sense